components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java

```diff
@@ -543,6 +569,39 @@ private void storeSessionData(AuthenticationContext context, String sessionConte
                         "user store domain: " + userStoreDomain + " in tenant domain: " + tenantDomain , e);
             }
         }
+
+        try {
+            if (appID != -1 && !UserSessionStore.getInstance().isExistingAppSession(sessionContextKey, subject,
+                    appID, appTenantId, inboundAuth)) {
+                UserSessionStore.getInstance().storeAppSessionData(sessionContextKey, subject,
+                        appID, appTenantId, inboundAuth);
+            }
+        } catch (DataAccessException e) {
+            throw new UserSessionException("Error while storing the App session data to the database.", e);
+        }
+    }
+
+    /**
+     * Method to store session meta data.
+     *
+     * @param sessionId Id of the authenticated session
+     * @param request   HttpServletRequest
+     * @throws UserSessionException while storing session meta data
+     */
+    private void storeSessionMetaData(String sessionId, HttpServletRequest request)
+            throws UserSessionException {
+        String userAgent = request.getHeader(javax.ws.rs.core.HttpHeaders.USER_AGENT);
+        String ip = request.getRemoteAddr();
+        String time = Long.toString(System.currentTimeMillis());
+
+        try {
+            UserSessionStore.getInstance().storeSessionMetaData(sessionId, "User Agent", userAgent);
```

Comment:
Need constants for these magic identifiers

Reply:
Done